sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroEncoder.ts

```diff
@@ -0,0 +1,217 @@
+} from "./models";
+import { SchemaDescription, SchemaRegistry } from "@azure/schema-registry";
+import { isMessageWithMetadata } from "./utility";
+
+interface CacheEntry {
+  /** Schema ID */
+  id: string;
+
+  /** avsc-specific representation for schema */
+  type: avro.Type;
+}
+
+const avroMimeType = "avro/binary";
+
+/**
+ * Avro encoder that obtains schemas from a schema registry and does not
+ * pack schemas into its payloads.
+ */
+export class SchemaRegistryAvroEncoder<MessageT = MessageWithMetadata> {
+  /**
+   * Creates a new encoder.
+   *
+   * @param client - Schema Registry where schemas are registered and obtained.
+   *                 Usually this is a SchemaRegistryClient instance.
+   */
+  constructor(client: SchemaRegistry, options?: SchemaRegistryAvroEncoderOptions<MessageT>) {
+    this.registry = client;
+    this.schemaGroup = options?.groupName;
+    this.autoRegisterSchemas = options?.autoRegisterSchemas ?? false;
+    this.messageAdapter = options?.messageAdapter;
+  }
+
+  private readonly schemaGroup?: string;
+  private readonly registry: SchemaRegistry;
+  private readonly autoRegisterSchemas: boolean;
+  private readonly messageAdapter?: MessageAdapter<MessageT>;
+
+  // REVIEW: signature.
+  //
+  // - Should we wrap all errors thrown by avsc to avoid having our exception //
+  //   contract being tied to its implementation details?
+  /**
+   * encodes the value parameter according to the input schema and creates a message
+   * with the encoded data.
+   *
+   * @param value - The value to encodeMessageData.
+   * @param schema - The Avro schema to use.
+   * @param options - Encoding options.
+   * @returns A new message with the encoded value. The structure of message is
+   * constrolled by the message factory option.
+   */
+  async encodeMessageData(value: unknown, schema: string): Promise<MessageT> {
+    const entry = await this.getSchemaByDefinition(schema);
+    const buffer = entry.type.toBuffer(value);
+    const payload = new Uint8Array(
+      buffer.buffer,
+      buffer.byteOffset,
+      buffer.byteLength / Uint8Array.BYTES_PER_ELEMENT
+    );
+    const contentType = `${avroMimeType}+${entry.id}`;
+    return this.messageAdapter
+      ? this.messageAdapter.produceMessage({
+          contentType,
+          body: payload,
+        })
+      : /**
+         * If no message consumer was provided, then a MessageWithMetadata will be
+         * returned. This should work because the MessageT type parameter defaults
+         * to MessageWithMetadata.
+         */
+        ({
+          body: payload,
+          contentType: contentType,
+        } as unknown as MessageT);
+  }
+
+  /**
+   * Decodes the payload of the message using the schema ID in the content type
+   * field if no schema was provided.
+   *
+   * @param message - The message with the payload to be decoded.
+   * @param options - Decoding options.
+   * @returns The decoded value.
+   */
+  async decodeMessageData(
```

Comment:
I think we need the preamble back compat support here.

Reply:
I wanted this PR to focus on the new encoder design only and was planning on adding the backward compat and the updates to the cache in separate PRs.

Reply:
Can we link to an issue?

Reply:
I created the following two issues for the February milestone:
[backward compat](https://github.com/Azure/azure-sdk-for-js/issues/20062)
[cache update](https://github.com/Azure/azure-sdk-for-js/issues/20064)